src/btree/bt_cursor.c

```diff
@@ -585,6 +586,32 @@ __cursor_search_neighboring(WT_CURSOR_BTREE *cbt, WT_CURFILE_STATE *state, int *
     if (F_ISSET(cursor, WT_CURSTD_PREFIX_SEARCH))
         return (ret);
 
+    /*
+     * We walked off the end of the tree. We need to check records before our original position to
+     * see if they are visible. In snapshot isolation we can restart our search from where we first
+     * searched as no records will appear in between.
+     *
+     * Due to the bounded cursor logic potentially changing the search key, we cannot rely on the
+     * saved state.
+     */
+    if (session->isolation == WT_ISO_SNAPSHOT && F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT) &&
+      !WT_CURSOR_BOUNDS_SET(cursor)) {
+        __cursor_state_restore(cursor, state);
+        if (btree->type == BTREE_ROW)
+            WT_RET(__cursor_row_search(cbt, true, NULL, NULL));
+        else
+            WT_RET(__cursor_col_search(cbt, NULL, NULL));
+
+        if (btree->type != BTREE_ROW)
```

Comment:
This part of the function could use some more comments explaining the decisions that are being made.